### PR TITLE
Documentation

### LINKED ISSUE #21: Move documentation into repo
To allow editing it during PRs

### PR DESCRIPTION
**Pull request by karottenreibe**
Fixes #21

- [x] Tests written
- [x] Documentation updated

### CODE REVIEW

**Comment by karottenreibe** [comment]
rfr

**Comment by karottenreibe** [comment]
You can download the rendered pdf from the dist here: https://ci.appveyor.com/project/cqse/teamscale-profiler-dotnet/build/18.2.1.44-bugvfpjr/artifacts

File: documentation/userguide.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+Once all the above-mentioned information is available and the Operations tasks have been completed, the Teamscale Administrator can create the dashboards:
+
+* Creation of one or more dashboards, depending on the test environment.
+* Customization of the dashboard, in order to collect the necessary data from the network share and start the preprocessing of the data.
+* For client applications: Creation of a Heartbeat dashboard and listing of all clients from which a regular Heartbeat trace is expected.
+* Setup of the exclusion criteria as parameters of the dashboard.
+
+
+# Client Heartbeat
```
Comment (agoeb) [inline documentation/userguide.md]: This is not part of the release, and we hardly ever use it anyway. Remove section?

File: documentation/userguide.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+* Facilitating access to the traces and binaries. These can either be made available on the test machine via a shared folder or transferred to a network share after test.
+* For client applications: installation of an Heartbeat application and activation of the profiler. Setup of a time-based execution. This enables a regular, automatic check of whether a client machine is able to send trace data to Teamscale.
+
+## Teamscale Administrator
+
+Once all the above-mentioned information is available and the Operations tasks have been completed, the Teamscale Administrator can create the dashboards:
+
+* Creation of one or more dashboards, depending on the test environment.
+* Customization of the dashboard, in order to collect the necessary data from the network share and start the preprocessing of the data.
+* For client applications: Creation of a Heartbeat dashboard and listing of all clients from which a regular Heartbeat trace is expected.
```
Comment (agoeb) [inline documentation/userguide.md]: See below: Remove heartbeat?

File: documentation/userguide.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+* the source code at the current revision
+* execution information, which is collected via the .NET Profiler
+
+The profiler writes method coverage information of a .NET application into a report file (a.k.a. *trace file*). It uses the standard .NET profiling interface and therefore can be used with any .NET application.
+
+The trace file is created immediately after the first call of a .NET method. However, it remains mostly empty as long as the process runs. This avoids an unnecessary performance overhead due to file accesses. As soon as the process ends gracefully, the report is finished and all collected information is written to the trace file.
+
+In the case of long running processes it must be taken into account that the report file is written only after the process has ended. In such a case, it must be ensured that the application is stopped and restarted in regular intervals or eager writing of the trace file is enabled.
```
Comment (agoeb) [inline documentation/userguide.md]: This completely ignores the "eagerness" mode. Update?

File: documentation/userguide.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+# Installation
+
+The .NET profiler supports both the .NET Standard and .NET Core runtime. For both variants the profiler has to be registered by environment variables before the application start. This can be achieved by either setting these variables in a startup script or by setting the variables globally. The former is preferred as otherwise all .NET aplications will be profiled.
+
+The default directory where traces are written to is `C:\Users\Public\`. See the Configuration section for specifying a custom output directory.
+
+The second part of this documentation handles special environments like applications running on IIS or Azure.
+
+## .NET Standard
```
Comment (agoeb) [inline documentation/userguide.md]: ".NET Standard" is a coined term in itself and extremely misleading. We mean ".NET Framework" here.

File: documentation/userguide.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+Please note that the profiler is **still** configured with variables with the `COR_PROFILER_` prefix. This especially holds for the `COR_PROFILER_TARGETDIR`.
+
+## Special Environments
+
+### IIS / Web Applications
+
+If the application is running in an IIS application server, the following steps need to be taken:
+
+* Make the relevant application pool see the environment variables
+* Regularly recycle the relevant application pool
```
Comment (agoeb) [inline documentation/userguide.md]: Completely ignoring eagerness here. Update?

File: documentation/userguide.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+IIS must be restarted after setting the environment variables in order for them to have any effect (e.g. by running iisreset in an administrator command prompt). On more recent IIS versions (tested with IIS 7), if you set the variables for the app pool account and not the service, recycling the application pool is enough and a full restart is not required.
+
+After restarting the IIS/recycling the pool and opening the application in a browser, a trace file should be created. As web applications are long running processes, it must be taken into account that the report file is written only after the process has ended, e.g. after stopping IIS, after stopping the profiled web site, or after recycling the application pool (cf. Section “When will the Report File be Written” TODO). In these cases, it must be ensured that the application is stopped and restarted in regular intervals. In the case of IIS, we suggest to adjust the configuration to automatically recycle the application pool every night.
```
Comment (agoeb) [inline documentation/userguide.md]: I spotted a "TODO" in this line. Do you haven an idea what that means?

Comment (agoeb) [reply documentation/userguide.md]: ...and again we only talk about recycling, but not about eagerness...

File: documentation/pdf.css [hunk 7]
```diff
@@ -0,0 +1,698 @@
...
+}
+
+body {
+  -ms-text-size-adjust: 100%;
+  -webkit-text-size-adjust: 100%;
+  line-height: 1.5;
+  color: #24292e;
+  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
```
Comment (agoeb) [inline documentation/pdf.css]: It seems AppVeyour is missing some fonts and somehow defaults to http://tholman.com/smelvetica/ - at least when the resulting PDF is viewed on macOS. I guess this selection only works good when it is interpreted on the machine it is viewed on, not when you generate a static PDF out of it and view that cross platform.

Comment (agoeb) [reply documentation/pdf.css]: Can we try and find a more reliable font for our use case?

File: documentation/userguide.md (hunk 3 above)
Comment (karottenreibe) [reply documentation/userguide.md]: The last part of the sentence is about eager mode

File: documentation/pdf.css (hunk 7 above)
Comment (karottenreibe) [reply documentation/pdf.css]: are you sure? It looks good on my machine:

selection_463

Here's the latest build: https://ci.appveyor.com/project/cqse/teamscale-profiler-dotnet/build/18.2.1.44-smgpshab/artifacts

**Comment by karottenreibe** [comment]
rfr again :-)

File: documentation/pdf.css (hunk 7 above)
Comment (agoeb) [reply documentation/pdf.css]: Looks indeed better on your machine compared to mine. It seems that fonts are even embedded into the PDF:
fonts
Unfortunately, it looks horrible for me: 
kerning
So I have no idea what we can do about it. Also: only the Seoge fonts looks so ugly, the Consolas text is fine.

**Pull request merged**